pyanaconda/modules/subscription/runtime.py

```diff
@@ -0,0 +1,111 @@
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+from pyanaconda.modules.common.task import Task
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+
+class SetRHSMConfigurationTask(Task):
+    """Task for setting configuration to the RHSM service.
+
+    Set configuration options of the RHSM service via it's
+    DBus interface, based on the provided SubscriptionRequest
+    structure.
+
+    Also in case one of the configuration options was unset,
+    restore the key to its original value. This way for example
+    a user decides at runtime to use the default server hostname
+    or RHSM baseurl, they can just delete the value in the UI,
+    triggering the original value to be restored when we encounter
+    the empty value for a key that originally was set to
+    a non empty value.
+    """
+
+    # Keys in the RHSM config key/value store we care about and
+    # should be able to restore to original value.
+    #
+    # NOTE: These keys map 1:1 to rhsm.conf. To see what they do the
+    #       best bet is to check the /etc/rhsm/rhsm.conf file on a system
+    #       with the subscription-manager package installed. The file
+    #       is heavily documented with comment's explaining what
+    #       the different keys do.
+    CONFIG_KEY_SERVER_HOSTNAME = "server.hostname"
+    CONFIG_KEY_SERVER_PROXY_HOSTNAME = "server.proxy_hostname"
+    CONFIG_KEY_SERVER_PROXY_PORT = "server.proxy_port"
+    CONFIG_KEY_SERVER_PROXY_USER = "server.proxy_user"
+    CONFIG_KEY_SERVER_PROXY_PASSWORD = "server.proxy_password"
+    CONFIG_KEY_RHSM_BASEURL = "rhsm.baseurl"
+
+    def __init__(self, rhsm_config_proxy, rhsm_config_defaults, subscription_request):
+        """Create a new task for setting RHSM configuration.
+
+        :param rhsm_config_proxy: DBus proxy for the RHSM Config object
+        :param dict rhsm_config_defaults: a dictionary of original RHSM configuration values
+        :param subscription_request: subscription request DBus
+        :type subscription_request: SubscriptionRequest instance
+        """
+        super().__init__()
+        self._rhsm_config_proxy = rhsm_config_proxy
```

Comment:
Due to issues with the RHSM system unit that provides the RHSM DBus API taking long to start, we have decided to pass all access to the RHSM DBus API through the `RHSMObserver` class instance, which makes sure attempts to get a proixy on 

We decided to pass all RHSM DBus API access via the `RHSMObserver`, which waits for the RHSM service to start before returning a proxy. Otherwise we could get crashes when we attempt to get a proxy for the RHSM DBus objects but the service is still starting.

Due to this we can't just call `get_proxy()` directly and need to go through `RHSMObserver`, which then the task needs to have access to. Rather then passing a reference to the Subscription module or the observer instance it seem like the best option to pass the proxy. This also simplifies testing.